### DISCUSSION TITLE
CDK CLI

### DISCUSSION THREAD

**Discussion by jamie-lynch** [guardian/cdk#181] [Show and tell]
The [CDK CLI](https://github.com/guardian/cdk-cli) is a tool we're working on in parallel to the `@guardian/cdk` library which aims to simplify some parts of working with CDK. We want to remove the tedious parts (boilerplate, linter config, etc.) and let you focus on defining your stacks. 

At the moment, the CLI offers three commands:

* **init** - This command can be used to generate the `cdk/` directory and all of the boilerplate required to get started from scratch including: `package.json`, linting config and build scripts.

* **new** - This command can be used to create a new stack within an existing project. You will be prompted for any parameter values which will be added to the file.

* **migrate** - This command can be used to migrate an existing cloudformation file into an existing cdk project. Currently, it can only migrate parameters but we hope to extend this to more resources in the future. 

**What's Next**

Below is a list of some of the proposed improvements and new features: 

* When running the new or migrate commands, also generate the required app file
* More specific output for either single- or multi-app projects
* Add test files alongside the stack and app files
* Create complete stack files from `@guardian/cdk` patterns
* Extend the migration command to handle more resources
* When running the migrate command, generate a snapshot test based on the cloudformation template
* Link the new and migrate commands with the init command to allow you to get started with a single command

If you've got any questions or any thoughts for improvements or features then let us know in the comments below!

**Comment by akash1810** [comment]
> Create complete stack files from @guardian/cdk patterns

This is super interesting. Should `cdk-cli` only ever output stacks that implement patterns (once we've written them)?

**Reply by jamie-lynch** [reply]
Yeah, I guess so. My thought was that you would provide a template name and any options either arguments or a series of prompts and it would spit out the entire stack all by magic. Ambitious but worth a try!

**Comment by davidfurey** [comment]
This looks really useful.  Would it be possible to add a command to automate the process of upgrading the boilerplate when it changes?

**Reply by jamie-lynch** [reply]
That's a nice idea! I'll have a think about how we might do that.

**Comment by jamie-lynch** [comment]
A quick update:

* When running the new or migrate commands, also generate the required app file - [Done](https://github.com/guardian/cdk-cli/pull/38)
* More specific output for either single- or multi-app projects - [Deprecated](https://github.com/guardian/cdk-cli/pull/49)
* Add test files alongside the stack and app files - [Done](https://github.com/guardian/cdk-cli/pull/63)
* Create complete stack files from @guardian/cdk patterns 
* Extend the migration command to handle more resources
* When running the migrate command, generate a snapshot test based on the cloudformation template
* Link the new and migrate commands with the init command to allow you to get started with a single command - [Done](https://github.com/guardian/cdk-cli/pull/66)
